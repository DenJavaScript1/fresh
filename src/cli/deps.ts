// std
export {
  join,
  resolve,
  toFileUrl,
} from "https://deno.land/std@0.128.0/path/mod.ts";
export { walk } from "https://deno.land/std@0.128.0/fs/walk.ts";
export { parse as parseArgs } from "https://deno.land/std@0.128.0/flags/mod.ts";
export {
  bold,
  green,
  red,
  yellow,
} from "https://deno.land/std@0.128.0/fmt/colors.ts";
