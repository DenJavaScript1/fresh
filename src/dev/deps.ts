// std
export {
  dirname,
  extname,
  fromFileUrl,
  join,
  resolve,
  toFileUrl,
} from "https://deno.land/std@0.128.0/path/mod.ts";
export { walk } from "https://deno.land/std@0.128.0/fs/walk.ts";
export { bold, red } from "https://deno.land/std@0.128.0/fmt/colors.ts";
export { parse } from "https://deno.land/std@0.128.0/flags/mod.ts";
