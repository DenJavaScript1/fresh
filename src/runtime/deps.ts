export * from "https://esm.sh/preact@10.5.14";
export {
  useCallback,
  useContext,
  useDebugValue,
  useEffect,
  useErrorBoundary,
  useImperativeHandle,
  useLayoutEffect,
  useMemo,
  useReducer,
  useRef,
  useState,
} from "https://esm.sh/preact@10.5.14/hooks";
export type {
  CreateHandle,
  EffectCallback,
  Inputs,
  PropRef,
  Reducer,
  Ref,
  StateUpdater,
} from "https://esm.sh/preact@10.5.14/hooks";
