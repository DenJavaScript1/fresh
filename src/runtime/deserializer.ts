// Run `deno run -A npm:esbuild --minify src/runtime/deserializer.ts` to minify
// this file. It is embedded into src/server/deserializer_code.ts.

export const KEY = "_f";

export function deserialize(str: string): unknown {
  function reviver(this: unknown, _key: string, value: unknown): unknown {
    if (typeof value === "object" && value && KEY in value) {
      // deno-lint-ignore no-explicit-any
      const v: any = value;
      if (v[KEY] === "l") {
        const val = v.v;
        val[KEY] = v.k;
        return val;
      }
      throw new Error(`Unknown key: ${v[KEY]}`);
    }
    return value;
  }

  const { v, r } = JSON.parse(str, reviver);
  const references = (r ?? []) as [string[], ...string[][]][];
  for (const [targetPath, ...refPaths] of references) {
    const target = targetPath.reduce((o, k) => k === null ? o : o[k], v);
    for (const refPath of refPaths) {
      if (refPath.length === 0) throw new Error("Invalid reference");
      // set the reference to the target object
      const parent = refPath.slice(0, -1).reduce(
        (o, k) => k === null ? o : o[k],
        v,
      );
      parent[refPath.at(-1)!] = target;
    }
  }
  return v;
}
