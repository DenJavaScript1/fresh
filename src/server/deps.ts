// -- std --
export {
  extname,
  fromFileUrl,
  toFileUrl,
} from "https://deno.land/std@0.177.0/path/mod.ts";
export { walk } from "https://deno.land/std@0.177.0/fs/walk.ts";
export { serve } from "https://deno.land/std@0.177.0/http/server.ts";
export type {
  ConnInfo,
  Handler as RequestHandler,
  ServeInit,
} from "https://deno.land/std@0.177.0/http/server.ts";
export { Status } from "https://deno.land/std@0.177.0/http/http_status.ts";
export {
  typeByExtension,
} from "https://deno.land/std@0.177.0/media_types/mod.ts";

// -- rutt --
export * as rutt from "https://deno.land/x/rutt@0.1.0/mod.ts";

// -- esbuild --
// @deno-types="https://deno.land/x/esbuild@v0.17.2/mod.d.ts"
import * as esbuildWasm from "https://deno.land/x/esbuild@v0.17.2/wasm.js";
import * as esbuildNative from "https://deno.land/x/esbuild@v0.17.2/mod.js";
// @ts-ignore trust me
const esbuild: typeof esbuildWasm = Deno.run === undefined
  ? esbuildWasm
  : esbuildNative;
export { esbuild, esbuildWasm as esbuildTypes };
// TODO(lino-levan): replace with tagged version
export { denoPlugin } from "https://raw.githubusercontent.com/lucacasonato/esbuild_deno_loader/b4b3ffb43e9380865ad12ab4d1c64c0ae0ca96cd/mod.ts";
