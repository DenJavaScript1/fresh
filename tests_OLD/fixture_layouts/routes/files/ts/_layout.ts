import { h } from "preact";
import type { PageProps } from "$fresh/server.ts";

export default function TsLayout({ Component }: PageProps) {
  return h("div", { class: "ts-layout" }, h(Component, null));
}
