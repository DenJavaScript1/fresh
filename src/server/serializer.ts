/**
 * This module contains a serializer for island props. The serializer is capable
 * of serializing the following:
 *
 * - `null`
 * - `boolean`
 * - `number`
 * - `string`
 * - `array`
 * - `object` (no prototypes)
 *
 * Circular references are supported and objects with the same reference are
 * serialized only once.
 *
 * The corresponding deserializer is in `src/runtime/deserializer.ts`.
 */
import { KEY } from "../runtime/deserializer.ts";

interface SerializeResult {
  /** The string serialization. */
  serialized: string;
  /** If the deserializer is required to deserialize this string. If this is
   * `false` the serialized string can be deserialized with `JSON.parse`. */
  requiresDeserializer: boolean;
}

export function serialize(data: unknown): SerializeResult {
  let requiresDeserializer = false;
  const seen = new Map<unknown, (string | null)[]>();
  const references = new Map<(string | null)[], (string | null)[][]>();

  const keyStack: (string | null)[] = [];
  const parentStack: unknown[] = [];

  let earlyReturn = false;

  const toSerialize = {
    v: data,
    get r() {
      earlyReturn = true;
      if (references.size > 0) {
        const refs = [];
        for (const [targetPath, refPaths] of references) {
          refs.push([targetPath, ...refPaths]);
        }
        return refs;
      }
      return undefined;
    },
  };

  function replacer(
    this: unknown,
    key: string | null,
    value: unknown,
  ): unknown {
    if (value === toSerialize || earlyReturn) {
      return value;
    }

    // For some object types, the path in the object graph from root is not the
    // same between the serialized representation, and deserialized objects. For
    // these cases, we have to change the contents of the key stack to match the
    // deserialized object.
    if (typeof this === "object" && this !== null && KEY in this) {
      if (this[KEY] === "l" && key === "v") key = null;
    }

    if (this !== toSerialize) {
      const parentIndex = parentStack.indexOf(this);
      parentStack.splice(parentIndex + 1);
      keyStack.splice(parentIndex);
      keyStack.push(key);
      // the parent is pushed before return
    }

    if (typeof value === "object" && value !== null) {
      const path = seen.get(value);
      const currentPath = [...keyStack];
      if (path !== undefined) {
        requiresDeserializer = true;
        const referenceArr = references.get(path);
        if (referenceArr === undefined) {
          references.set(path, [currentPath]);
        } else {
          referenceArr.push(currentPath);
        }
        return 0;
      } else {
        seen.set(value, currentPath);
      }
    }

    if (typeof value === "object" && value && KEY in value) {
      requiresDeserializer = true;
      // deno-lint-ignore no-explicit-any
      const v: any = { ...value };
      const k = v[KEY];
      delete v[KEY];
      const res = { [KEY]: "l", k, v };
      parentStack.push(res);
      return res;
    } else {
      parentStack.push(value);
      return value;
    }
  }

  const serialized = JSON.stringify(toSerialize, replacer);
  return { serialized, requiresDeserializer };
}
