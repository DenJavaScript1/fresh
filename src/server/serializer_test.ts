// deno-lint-ignore-file no-explicit-any

import { serialize } from "./serializer.ts";
import { assert, assertEquals, assertSnapshot } from "../../tests/deps.ts";
import { deserialize, KEY } from "../runtime/deserializer.ts";

Deno.test("serializer - primitives & plain objects", async (t) => {
  const data = {
    a: 1,
    b: "2",
    c: true,
    d: null,
    f: [1, 2, 3],
    g: { a: 1, b: 2, c: 3 },
  };
  const res = serialize(data);
  assert(!res.requiresDeserializer);
  await assertSnapshot(t, res.serialized);
  const deserialized = deserialize(res.serialized);
  assertEquals(deserialized, data);
});

Deno.test("serializer - magic key", async (t) => {
  const data = { [KEY]: "f", a: 1 };
  const res = serialize(data);
  assert(res.requiresDeserializer);
  await assertSnapshot(t, res.serialized);
  const deserialized = deserialize(res.serialized);
  assertEquals(deserialized, data);
});

Deno.test("serializer - circular reference objects", async (t) => {
  const data: any = { a: 1 };
  data.b = data;
  const res = serialize(data);
  assert(res.requiresDeserializer);
  await assertSnapshot(t, res.serialized);
  const deserialized = deserialize(res.serialized);
  assertEquals(deserialized, data);
});

Deno.test("serializer - circular reference nested objects", async (t) => {
  const data: any = { a: 1, b: { c: 2 } };
  data.b.d = data;
  const res = serialize(data);
  assert(res.requiresDeserializer);
  await assertSnapshot(t, res.serialized);
  const deserialized = deserialize(res.serialized);
  assertEquals(deserialized, data);
});

Deno.test("serializer - circular reference array", async (t) => {
  const data: any = [1, 2, 3];
  data.push(data);
  const res = serialize(data);
  assert(res.requiresDeserializer);
  await assertSnapshot(t, res.serialized);
  const deserialized: any = deserialize(res.serialized);
  assertEquals(deserialized, data);
  assertEquals(deserialized.length, 4);
});

Deno.test("serializer - multiple reference", async (t) => {
  const data: any = { a: 1, b: { c: 2 } };
  data.d = data.b;
  const res = serialize(data);
  assert(res.requiresDeserializer);
  await assertSnapshot(t, res.serialized);
  const deserialized = deserialize(res.serialized);
  assertEquals(deserialized, data);
});

Deno.test("serializer - multiple reference in magic key", async (t) => {
  const inner = { foo: "bar" };
  const literal: any = { [KEY]: "x", inner };
  const data = { literal, inner };
  const res = serialize(data);
  assert(res.requiresDeserializer);
  await assertSnapshot(t, res.serialized);
  const deserialized: any = deserialize(res.serialized);
  assertEquals(deserialized, data);
});
