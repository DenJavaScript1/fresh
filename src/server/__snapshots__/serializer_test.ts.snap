export const snapshot = {};

snapshot[`serializer - primitives & plain objects 1`] = `'{"v":{"a":1,"b":"2","c":true,"d":null,"f":[1,2,3],"g":{"a":1,"b":2,"c":3}}}'`;

snapshot[`serializer - magic key 1`] = `'{"v":{"_f":"l","k":"f","v":{"a":1}}}'`;

snapshot[`serializer - circular reference objects 1`] = `'{"v":{"a":1,"b":0},"r":[[[],["b"]]]}'`;

snapshot[`serializer - circular reference nested objects 1`] = `'{"v":{"a":1,"b":{"c":2,"d":0}},"r":[[[],["b","d"]]]}'`;

snapshot[`serializer - circular reference array 1`] = `'{"v":[1,2,3,0],"r":[[[],["3"]]]}'`;

snapshot[`serializer - multiple reference 1`] = `'{"v":{"a":1,"b":{"c":2},"d":0},"r":[[["b"],["d"]]]}'`;

snapshot[`serializer - multiple reference in magic key 1`] = `'{"v":{"literal":{"_f":"l","k":"x","v":{"inner":{"foo":"bar"}}},"inner":0},"r":[[["literal",null,"inner"],["inner"]]]}'`;
